src/cluster/placement/placements_watcher.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package placement
+
+import (
+	"errors"
+	"sync"
+
+	"go.uber.org/atomic"
+
+	"github.com/m3db/m3/src/cluster/generated/proto/placementpb"
+	"github.com/m3db/m3/src/cluster/kv"
+	"github.com/m3db/m3/src/cluster/kv/util/runtime"
+)
+
+var (
+	errNilValue             = errors.New("nil value received")
+	errWatcherIsNotWatching = errors.New("placement watcher is not watching")
+	errWatcherIsWatching    = errors.New("placement watcher is watching")
+	errWatcherCastError     = errors.New("interface cast failed, unexpected placement type")
+
+	_ Watcher = (*placementsWatcher)(nil) // enforce interface compliance
+)
+
+// placementsWatcher implements watcher of staged placement.
+// TODO: Consider migrating to non-staged placement.
+// Currently, the aggregator placement is stored as staged placement in etcd using
+// protobuf type `*placementpb.PlacementSnapshots` as a remnant of now deprecated concept
+// of "staged placement". This Watcher abstracts that detail from the clients while
+// maintaining backward compatibility.
+// Migrating from storing `*placementpb.PlacementSnapshots` to storing `*placementpb.Placement`
+// in etcd is desirable in future, although not necessary.
+// It would have to be done in a graceful manner that doesn't break clients,
+// such as in the following sequence:
+// - change code that manages placement to store both types of placement protobufs in different
+//   well-known etcd keys within a single transaction
+// - deploy that code and make at least one placement change
+// - change this watcher code to read the non-staged placement from respective etcd key
+// - upgrade all deployed clients (this is most difficult as we don't control all the deployments
+//   out there in the world)
+// - modify placement management code to stop storing staged placement
```

Comment:
Can the sequence be:
1, create a watcher or a placement type that is able to unmarshal both `*placementpb.PlacementSnapshots` and `*placementpb.Placement`
2, deploy collector and aggregator to this version
3, make placement change

Reply:
Good question. How would it be able to unmarshal both without knowing what type to expect? Do you mean "try unmarshal one type, if fails - try the other" type of approach? Not sure if that'd good enough.

Reply:
Yeah I'm thinking about the approach to try one type and then another

Reply:
It may work if migration is done carefully, but if something goes wrong - it may be harder to debug. Anyways, let me remove those details from the comment so it doesn't appear like there's only one option to do such migration.

Reply:
I'll create an issue for it